command/var_put.go

```diff
@@ -0,0 +1,472 @@
+	var err error
+	var forceOverwrite, enforce, doVerbose bool
+	var path, checkIndexStr string
+	var checkIndex uint64
+
+	flags := c.Meta.FlagSet(c.Name(), FlagSetClient)
+	flags.Usage = func() { c.Ui.Output(c.Help()) }
+
+	flags.StringVar(&c.inFmt, "in", "json", "")
+	flags.StringVar(&checkIndexStr, "check-index", "", "")
+	flags.BoolVar(&forceOverwrite, "force", false, "")
+	flags.BoolVar(&doVerbose, "v", false, "")
+	flags.StringVar(&c.tmpl, "template", "", "")
+	if fileInfo, _ := os.Stdout.Stat(); (fileInfo.Mode() & os.ModeCharDevice) != 0 {
+		flags.StringVar(&c.outFmt, "out", "none", "")
+	} else {
+		flags.StringVar(&c.outFmt, "out", "json", "")
+	}
+
+	if err := flags.Parse(args); err != nil {
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	args = flags.Args()
+
+	// Manage verbose output
+	verbose := func(_ string) {} //no-op
+	if doVerbose {
+		verbose = func(msg string) {
+			c.Ui.Warn(msg)
+		}
+	}
+	// Parse the check-index
+	checkIndex, enforce, err = parseCheckIndex(checkIndexStr)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error parsing check-index value %q: %v", checkIndexStr, err))
+		return 1
+	}
+
+	// Pull our fake stdin if needed
+	stdin := (io.Reader)(os.Stdin)
+	if c.testStdin != nil {
+		stdin = c.testStdin
+	}
+
+	switch {
+	case len(args) < 1:
+		c.Ui.Error(fmt.Sprintf("Not enough arguments (expected >1, got %d)", len(args)))
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	case len(args) == 1 && !isArgStdinRef(args[0]) && !isArgFileRef(args[0]):
+		c.Ui.Error("Must supply data")
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	if err = c.validateInputFlag(); err != nil {
+		c.Ui.Error(err.Error())
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	if err := c.validateOutputFlag(); err != nil {
+		c.Ui.Error(err.Error())
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	arg := args[0]
+	switch {
+	// Handle first argument: can be -, @file, «var path»
+	case isArgStdinRef(arg):
+
+		// read the specification into memory from stdin
+		stat, _ := os.Stdin.Stat()
+		if (stat.Mode() & os.ModeCharDevice) == 0 {
+			c.contents, err = io.ReadAll(os.Stdin)
+			if err != nil {
+				c.Ui.Error(fmt.Sprintf("Error reading from stdin: %s", err))
+				return 1
+			}
+		}
+		verbose(fmt.Sprintf("Reading whole %s variable specification from stdin", strings.ToUpper(c.inFmt)))
+
+	case isArgFileRef(arg):
+		// ArgFileRefs start with "@" so we need to peel that off
+		filepath := arg[1:]
+		// detect format based on extension
+		p := strings.Split(filepath, ".")
+		if len(p) < 2 {
+			c.Ui.Error(fmt.Sprintf("Unable to determine format of %s; Use the -in flag to specify it.", filepath))
+			return 1
+		}
+		// detect format based on file extension
+		switch strings.ToLower(p[len(p)-1]) {
+		case "json":
+			c.inFmt = "json"
+		case "hcl":
+			c.inFmt = "hcl"
+		default:
+			c.Ui.Error(fmt.Sprintf("Unable to determine format of %s; Use the -in flag to specify it.", filepath))
+			return 1
+		}
```

Comment:
There's a nice helper in the stdlib for this:

```suggestion
		// detect format based on file extension
		fpath := arg[1:]
		switch filepath.Ext(fpath) {
		case "json":
			c.inFmt = "json"
		case "hcl":
			c.inFmt = "hcl"
		default:
			c.Ui.Error(fmt.Sprintf("Unable to determine format of %s; Use the -in flag to specify it.", filepath))
			return 1
		}
```